mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala

```diff
@@ -224,142 +261,133 @@ class KMeans private (
 
-      bcActiveCenters.unpersist(blocking = false)
-
-      // Update the cluster centers and costs for each active run
-      for ((run, i) <- activeRuns.zipWithIndex) {
-        var changed = false
-        var j = 0
-        while (j < k) {
-          val (sum, count) = totalContribs((i, j))
-          if (count != 0) {
-            scal(1.0 / count, sum)
-            val newCenter = new VectorWithNorm(sum)
-            if (KMeans.fastSquaredDistance(newCenter, centers(run)(j)) > epsilon * epsilon) {
-              changed = true
-            }
-            centers(run)(j) = newCenter
+
+      bcCenters.unpersist(blocking = false)
+
+      var changed = false
+      var j = 0
+      while (j < k) {
+        val (sum, count) = totalContribs(j)
+        if (count != 0) {
+          scal(1.0 / count, sum)
+          val newCenter = new VectorWithNorm(sum)
+          if (KMeans.fastSquaredDistance(newCenter, centers(j)) > epsilon * epsilon) {
+            changed = true
           }
-          j += 1
-        }
-        if (!changed) {
-          active(run) = false
-          logInfo("Run " + run + " finished in " + (iteration + 1) + " iterations")
+          centers(j) = newCenter
         }
-        costs(run) = costAccums(i).value
+        j += 1
       }
 
-      activeRuns = activeRuns.filter(active(_))
+      costs = costAccums.value
+      if (!changed) {
+        logInfo("Run finished in " + (iteration + 1) + " iterations")
+        done = true
+      }
       iteration += 1
     }
 
     val iterationTimeInSeconds = (System.nanoTime() - iterationStartTime) / 1e9
     logInfo(s"Iterations took " + "%.3f".format(iterationTimeInSeconds) + " seconds.")
-
     if (iteration == maxIterations) {
       logInfo(s"KMeans reached the max number of iterations: $maxIterations.")
     } else {
       logInfo(s"KMeans converged in $iteration iterations.")
     }
+    logInfo(s"The cost is $costs.")
 
-    val (minCost, bestRun) = costs.zipWithIndex.min
-
-    logInfo(s"The cost for the best run is $minCost.")
-
-    new KMeansModel(centers(bestRun).map(_.vector))
+    new KMeansModel(centers.map(_.vector))
   }
 
   /**
-   * Initialize `runs` sets of cluster centers at random.
+   * Initialize cluster centers at random.
    */
-  private def initRandom(data: RDD[VectorWithNorm])
-  : Array[Array[VectorWithNorm]] = {
-    // Sample all the cluster centers in one pass to avoid repeated scans
-    val sample = data.takeSample(true, runs * k, new XORShiftRandom(this.seed).nextInt()).toSeq
-    Array.tabulate(runs)(r => sample.slice(r * k, (r + 1) * k).map { v =>
-      new VectorWithNorm(Vectors.dense(v.vector.toArray), v.norm)
-    }.toArray)
+  private def initRandom(data: RDD[VectorWithNorm]): Array[VectorWithNorm] = {
+    // Sample cluster centers in one pass
+    val sample = data.takeSample(true, k, new XORShiftRandom(this.seed).nextInt()).toSeq
+    sample.map { v => new VectorWithNorm(Vectors.dense(v.vector.toArray), v.norm) }.toArray
```

Comment:
Unnecessary copy with `.map`. Can be avoided with pre-allocated array and transformation made within loop.